File: sandbox-runner/patterns/early-return/python/starter.py
import asyncio
import time

from temporalio.client import (
    Client,
    WithStartWorkflowOperation,
    WorkflowUpdateStage,
)
from temporalio.common import WorkflowIDConflictPolicy

from shared import TASK_QUEUE, WORKFLOW_ID_PREFIX, TransactionRequest
from workflows import TransactionWorkflow


async def main() -> None:
    client = await Client.connect("localhost:7233")
    transaction_id = f"{WORKFLOW_ID_PREFIX}-{int(time.time() * 1000)}"
    req = TransactionRequest(amount=100, currency="USD")

    start_op = WithStartWorkflowOperation(
        TransactionWorkflow.run,
        req,
        id=transaction_id,
        task_queue=TASK_QUEUE,
        id_conflict_policy=WorkflowIDConflictPolicy.FAIL,
    )

    t0 = time.monotonic()
    update_handle = await client.start_update_with_start_workflow(
        TransactionWorkflow.return_init_result,
        wait_for_stage=WorkflowUpdateStage.COMPLETED,
        start_workflow_operation=start_op,
    )
    tx = await update_handle.result()
    elapsed_ms = int((time.monotonic() - t0) * 1000)
    print(f"Early return after {elapsed_ms}ms: {tx.id} (status={tx.status})")

    handle = client.get_workflow_handle(transaction_id)
    final = await handle.result()
    elapsed_ms = int((time.monotonic() - t0) * 1000)
    print(f"Workflow completed after {elapsed_ms}ms: {final}")
    print(
        f"Open the Temporal UI and search for '{transaction_id}' to see the history."
    )


if __name__ == "__main__":
    asyncio.run(main())

File: sandbox-runner/patterns/early-return/typescript/starter.ts
import { Client, Connection, WithStartWorkflowOperation } from "@temporalio/client";

import { TASK_QUEUE, WORKFLOW_ID_PREFIX, type TransactionRequest } from "./shared";
import { returnInitResultUpdate, transactionWorkflow } from "./workflows";

async function main(): Promise<void> {
  const connection = await Connection.connect();
  try {
    const client = new Client({ connection });
    const transactionId = `${WORKFLOW_ID_PREFIX}-${Date.now()}`;
    const req: TransactionRequest = { amount: 100, currency: "USD" };

    const startOp = new WithStartWorkflowOperation(transactionWorkflow, {
      workflowId: transactionId,
      args: [req],
      taskQueue: TASK_QUEUE,
      workflowIdConflictPolicy: "FAIL",
    });

    const t0 = Date.now();
    const tx = await client.workflow.executeUpdateWithStart(returnInitResultUpdate, {
      startWorkflowOperation: startOp,
    });
    console.log(
      `Early return after ${Date.now() - t0}ms: ${tx.id} (status=${tx.status})`,
    );

    const handle = await startOp.workflowHandle();
    const final = await handle.result();
    console.log(`Workflow completed after ${Date.now() - t0}ms: ${JSON.stringify(final)}`);
    console.log(
      `Open the Temporal UI and search for '${transactionId}' to see the history.`,
    );
  } finally {
    await connection.close();
  }
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
